src/ngraph/op/concat.cpp

```diff
@@ -46,12 +46,16 @@ void op::Concat::validate_and_infer_types()
 
     for (uint64_t i = 0; i < get_input_size(); i++)
     {
+        NODE_VALIDATION_CHECK(this,
+                              element::Type::merge(inputs_et, inputs_et, get_input_element_type(i)),
+                              "Argument element types are inconsistent.");
         PartialShape this_input_shape = get_input_partial_shape(i);
         Dimension this_input_rank = this_input_shape.rank();
         if (this_input_rank.is_static())
         {
+            m_axis = m_axis < 0 ? m_axis + int64_t(this_input_rank) : m_axis;
```

Comment:
Don't update m_axis. This will not work for dynamic shape.
Example) m_axis originally is -1 and Rank is 3. Then your code will update m_axis to 2. Then let's say on the next call Rank becomes 4, since m_axis was -1 originally the correct concat axis will be 3. But your code has updated m_axis to 2 and you get an incorrect concat axis.
For node validation, assign the computed concat axis to a new variable like concat_axis and use it in the NODE_VALIDATION_CHECK below.

Reply:
Agree. You can provide a size_t get_axis() and int64_t get_axis(). size_t get_axis() will return the adjusted axes and throws an error if input shape is not static. int64_t get_axis() will return the value specified through the constructors and setters.

Reply:
If axis will not be updated in this way, then the value of axis is not correct...  Axis and rank are defined as follows: {axis = -1, Rank = 2} thus the axis is equal 1.  For a reference version we have { axis = -1 Rank = 3} axis = 2. But we need to have 1